File: 0046/main.py
class Solution:
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """

        res = [[]]
        for num, x in enumerate(nums):
            temp = res.copy()
            res = list()
            for L in temp:
                res = [L[:i]+[x]+L[i:] for i in range(num+1)]
        return res

File: 0046/main.cpp
class Solution {
private:
    vector<vector<int>> res;
    void helper(vector<int> nums, int n) {
        if(n == nums.size()-1) return;
        helper(nums, n+1);
        for(int i=n+1;i<nums.size();++i) {
            res.push_back(nums);
            swap(res.back()[n], res.back()[i]);
            helper(res.back(), n+1);
        }
    }
public:
    vector<vector<int>> permute(vector<int>& nums) {
        res.push_back(nums);
        if(nums.empty()) return res;
        helper(nums, 0);
        return res;
    }
};